logging/logger.go

```diff
@@ -0,0 +1,121 @@
+// MIT License. Copyright (c) 2020 CQFN
+// https://github.com/cqfn/degitx/blob/master/LICENSE
+
+// Package logging contains degitx logging API, mapping and configuring internal implementation.
+// zap is current implementation
+package logging
+
+import (
+	"os"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type LogLevel int8
+
+const (
+	Debug LogLevel = 0
+	Info  LogLevel = 1
+	Warn  LogLevel = 2
+	Error LogLevel = 3
+)
+
+type Logger struct {
+	internalLogger        *zap.Logger
+	internalSugaredLogger *zap.SugaredLogger
+}
+
+func NewLogger(tag string) (*Logger, error) {
+	internalLogger, internalSugaredLogger, err := initInternalLogger(tag)
+	if err != nil {
+		return nil, err
+	}
+	return &Logger{
+			internalLogger,
+			internalSugaredLogger,
+		},
+		nil
+}
+
+func initInternalLogger(tag string) (*zap.Logger, *zap.SugaredLogger, error) {
+	consoleDebugging := zapcore.Lock(os.Stdout)
+	consoleErrors := zapcore.Lock(os.Stderr)
+
+	spec := []zapcore.Field{
+		zap.String("tag", tag),
+		zap.String("locator", locator),
+	}
+
+	sink, errSink, err := openSinks()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	core := zapcore.NewTee(
+		zapcore.NewCore(consoleEncoder, consoleErrors, highPriority).With(spec),
+		zapcore.NewCore(consoleEncoder, consoleDebugging, lowPriority).With(spec),
+		zapcore.NewCore(jsonEncoder, sink, highPriority).With(spec),
+	)
+
+	logger := zap.New(
+		core,
+		zap.ErrorOutput(errSink),
+		zap.AddStacktrace(zapcore.ErrorLevel))
+	defer logger.Sync() //nolint:errcheck // zap docs says it normal way to use Sync()
+	suggared := logger.Sugar()
+	return logger, suggared, nil
+}
+
+func SetLevel(logLevel LogLevel) {
+	logLvl.SetLevel(internalLogLevel[logLevel])
+}
+
+func (log *Logger) Debug(template string) {
+	log.internalLogger.Debug(template)
+}
+
+func (log *Logger) Info(template string) {
+	log.internalLogger.Info(template)
+}
+
+func (log *Logger) Warn(template string) {
+	log.internalLogger.Warn(template)
+}
+
+func (log *Logger) Error(template string) {
+	log.internalLogger.Error(template)
+}
+
+func (log *Logger) Check(level LogLevel, s string) interface{} {
+	return log.internalLogger.Check(internalLogLevel[level], s)
+}
+
+func (log *Logger) Debugf(template string, args ...interface{}) {
+	log.internalSugaredLogger.Debugf(template, args...)
+}
+
+func (log *Logger) Infof(template string, args ...interface{}) {
+	log.internalSugaredLogger.Infof(template, args...)
+}
+
+func (log *Logger) Warnf(template string, args ...interface{}) {
+	log.internalSugaredLogger.Warnf(template, args...)
+}
+
+func (log *Logger) Errorf(template string, args ...interface{}) {
+	log.internalSugaredLogger.Errorf(template, args...)
+}
+
+func openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error) {
+	sink, closeOut, err := zap.Open("./errors")
+	if err != nil {
+		return nil, nil, err
+	}
+	errSink, _, err := zap.Open("./logInternalErrors")
```

Comment:
@stepanvalyavskiy what is the difference between `errors` and `logInternalErrors`? Can we use one file for both?

Reply:
yes we can, but `logInternalErrors` is specific for Zap logger. `errSink` is an output for internals zap errors, that wasn't logged by users of this lib.